app/assets/v2/js/grants/cart/algorand_extension.js

```diff
@@ -0,0 +1,208 @@
+
+      balance.assets.map(asset => {
+        if (asset['asset-id'] == token.addr && asset['amount'] <= amount * 10 ** token.decimals)
+          has_enough_asset_balance = true;
+      });
+
+      if (has_enough_asset_balance) {
+        _alert({ message: `Insufficent balance in address ${from_address}` }, 'danger');
+        return;
+      }
+      
+      // step4: set modal to waiting state
+      vm.updatePaymentStatus(grant.grant_id, 'waiting');
+
+      // step5: get txnParams
+      const txParams = await AlgoSigner.algod({
+        ledger: NETWORK,
+        path: '/v2/transactions/params'
+      });
+
+      let txn;
+      // step5: sign transaction
+
+      if (token_name == 'ALGO') {
+        // ALGO token
+        txn = {
+          from: from_address.toUpperCase(),
+          to: to_address.toUpperCase(),
+          fee: txParams['fee'],
+          type: 'pay',
+          amount: amount * 10 ** token.decimals,
+          firstRound: txParams['last-round'],
+          lastRound: txParams['last-round'] + 1000,
+          genesisID: txParams['genesis-id'],
+          genesisHash: txParams['genesis-hash'],
+          note: 'contributing to gitcoin grant'
+        };
+      } else {
+        // ALGO assets
+        txn = {
+          from: from_address.toUpperCase(),
+          to: to_address.toUpperCase(),
+          assetIndex: Number(token.addr),
+          note: 'contributing to gitcoin grant',
+          amount: amount * 10 ** token.decimals,
+          type: 'axfer',
+          fee: txParams['min-fee'],
+          firstRound: txParams['last-round'],
+          lastRound: txParams['last-round'] + 1000,
+          genesisID: txParams['genesis-id'],
+          genesisHash: txParams['genesis-hash']
+        };
+      }
+
+      AlgoSigner.sign(txn).then(signedTx => {
+        // step7: broadcast txn
+        AlgoSigner.send({
+          ledger: NETWORK,
+          tx: signedTx.blob
+        }).then(tx => {
+          callback(null, from_address, tx.txId);
+        }).catch((e) => {
+          console.log(e);
+          _alert({ message: 'Unable to broadcast txn. Please try again' }, 'danger');
+          vm.updatePaymentStatus(grant.grant_id, 'failed');
+          return;
+        });
+
+      }).catch(e => {
+        console.log(e);
+        _alert({ message: 'Unable to sign txn. Please try again' }, 'danger');
+        vm.updatePaymentStatus(grant.grant_id, 'failed');
+        return;
+      });
+
+    }).catch(e => {
+      console.log(e);
+      _alert({ message: 'Please allow Gitcoin to connect to AlgoSigner extension' }, 'danger');
+      vm.updatePaymentStatus(grant.grant_id, 'failed');
+      return;
+    });
+  } catch (e) {
+    callback(err);
+    return;
+  }
+  
+
+  function callback(error, from_address, txn) {
+    if (error) {
+      vm.updatePaymentStatus(grant.grant_id, 'failed');
+      _alert({ message: gettext('Unable to contribute to grant due to ' + error) }, 'danger');
```

Comment:
do we have to append `error`? is it a non-technical message?

Reply:
Form the flows I've tested. Yes